gcs_contents_manager.py

```diff
@@ -428,3 +449,262 @@ def rename_file(self, old_path, new_path):
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def save(self, model, path):
+    if path in ['', '/']:
+      raise HTTPError(403, 'The top-level directory is read-only')
+    try:
+      self.run_pre_save_hook(model=model, path=path)
+
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
+      if relative_path in ['', '/']:
+        raise HTTPError(403, 'The top-level directory contents are read-only')
+
+      if 'path' in model:
+        model['path'] = relative_path
+
+      model = cm.save(model, relative_path)
+      if 'path' in model:
+        model['path'] = path
+      return model
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def delete_file(self, path):
+    if path in ['', '/']:
+      raise HTTPError(403, 'The top-level directory is read-only')
+    try:
+      cm, relative_path, path_prefix = self._content_manager(path)
+      if not cm:
+        raise HTTPError(404, 'Not Found')
+      if relative_path in ['', '/']:
+        raise HTTPError(403, 'The top-level directory contents are read-only')
+      return cm.delete_file(relative_path)
+    except OSError as err:
+      # The built-in file contents manager will not attempt to wrap permissions
+      # errors when deleting files if they occur while trying to move the
+      # to-be-deleted file to the trash, because the underlying send2trash
+      # library does not set the errno attribute of the raised OSError.
+      #
+      # To work around this we explicitly catch such errors, check if they
+      # start with the magic text "Permission denied", and then wrap them
+      # in an HTTPError.
+      if str(err).startswith('Permission denied'):
+        raise HTTPError(403, str(err))
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(err.errno, str(err)))
+    except HTTPError as err:
+      raise err
+    except Exception as ex:
+      raise HTTPError(500, 'Internal server error: [{}] {}'.format(type(ex), str(ex)))
+
+  def rename_file(self, old_path, new_path):
+    if old_path in ['', '/']:
+      raise HTTPError(403, 'The top-level directory is read-only')
+    try:
+      old_cm, old_relative_path, _ = self._content_manager(old_path)
+      if not old_cm:
+        raise HTTPError(404, 'Not Found')
+      if old_relative_path in ['', '/']:
+        raise HTTPError(403, 'The top-level directory contents are read-only')
+
+      new_cm, new_relative_path, _ = self._content_manager(new_path)
+      if not new_cm:
+        raise HTTPError(404, 'Not Found')
+      if new_relative_path in ['', '/']:
+        raise HTTPError(403, 'The top-level directory contents are read-only')
+
+      if old_cm != new_cm:
+        raise HTTPError(400, 'Bad Request')
```

Comment:
Similarly, this could have a better error message noting that you can't rename across file systems

Reply:
How difficult would be to support renames/moves across file systems? It seems like a pretty useful feature to e.g move my local notebooks into GCS and vice versa from the UI.

Reply:
@JerryLeiDing It would probably be feasible to support renames across file systems, but that is a large enough effort that I would want to do it in its own pull request, and only do it after we have some form of integration tests.